File: apps/backend/src/modules/identity/handler.ts
import type { AuthenticationMiddleware } from "../../middleware/auth.ts";
import type { IIdentitySvc } from "./svc.ts";
import { type Request, type Response, Router } from "express";
import type { Logger } from "winston";

export class IdentityHandler {
  private svc: IIdentitySvc;
  private authenticationMiddleware: AuthenticationMiddleware;
  private logger: Logger;
  public router: Router;
  constructor(
    svc: IIdentitySvc,
    authenticationMiddleware: AuthenticationMiddleware,
    logger: Logger,
  ) {
    this.svc = svc;
    this.authenticationMiddleware = authenticationMiddleware;
    this.router = Router();
    this.logger = logger.child({
      service: "identity-handler",
    });
    this.registerRoutes();
  }

  private registerRoutes = () => {
    this.router.get(
      "/properties/:propertyId/users/:userId",
      this.authenticationMiddleware,
      async (req: Request, res: Response) => {
        const { propertyId, userId } = req.params;
        const { data, error } = await this.svc.getPropertyUser(
          propertyId,
          userId,
        );
        if (error) {
          this.logger.error(error.message, {
            route: req.route?.path,
            method: req.method,
            statusCode: error.code,
          });
          res.status(error.code).json({ error: error });
          return;
        }
        res.status(200).json({ data: data });
      },
    );
  };
}

File: apps/backend/src/modules/identity/handler.py
from uuid import UUID
from fastapi import APIRouter, Depends, status

from src.modules.identity.svc import IIdentitySvc
from src.shared.types.modules.identity.model import PropertyUser
from src.shared.types.response import (
    AuthMiddlewareResponse,
    AuthMiddlewareSignature,
    HTTPError,
)


class IdentityHandler:
    def __init__(
        self,
        svc: IIdentitySvc,
        auth_middleware: AuthMiddlewareSignature,
    ):
        self.svc = svc
        self.router = APIRouter()
        self.auth_middleware = auth_middleware
        self.register_routes()

    def register_routes(self):
        @self.router.get(
            "/properties/{property_id}/users/{user_id}",
            status_code=status.HTTP_200_OK,
            responses={
                200: {"model": PropertyUser},
                "default": {"model": HTTPError},
            },
        )
        async def get_property_user(
            property_id: UUID,
            user_id: UUID,
            user_result: AuthMiddlewareResponse = Depends(self.auth_middleware),
        ) -> PropertyUser:
            return await self.svc.get_property_user(
                property_id=property_id, user_id=user_id
            )
